### PR TITLE
Add Protractor and Intern.js comparison

### PR DESCRIPTION
**Pull request by ionut998**

### CODE REVIEW

**Comment by reevoo-samuel** [comment]
*If you're new, read the [guidelines]
(https://github.com/reevoo/guidelines/blob/master/pull_requests.md).*

Good code is **tested**, **designed for change**, **easy to understand**
and **able to cope when things go wrong**.

Good pull requests are **descriptive**, **small** and **short-lived**.

**For the Author**
- [ ] Description explains what you have done and why.
- [ ] Tests have been written (or you have explained why you don't need
tests).
- [ ] Documentation/comments have been added/updated.

**For the Reviewer**
- [ ] The code is well-designed and able to cope with change.
- [ ] The code is easy to understand.
- [ ] Tests cover happy and unhappy paths in the code.
- [ ] Tests are green on CI (or pulled/tested on your machine).

File: _posts/2016-06-08-protractor-vs-intern-js.md [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+layout: post
+title:  Protractor or Intern ?
+date:   2016-06-08 14:00
+categories: blog
+navId: blog
+author: Alexandru Anca
+github: ionut998
+---
+
+The focus of this post is a comparison between protractor and intern for the integration testing of an angular application.
```
Comment (irikuo) [inline _posts/2016-06-08-protractor-vs-intern-js.md]: The name of Protractor, Intern and AngluarJS should match the official naming for consistency.

File: _posts/2016-06-08-protractor-vs-intern-js.md [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+
+# Current Implementation
+
+Currently the application is tested using Ruby and Capybara framework. As it turns out using this combination to test a single page application is very flaky. Out of 3 CI runs only 1 is usually successful, the rest is the result of flakiness. 
+Another down side of this is that you need to install ruby on top of your application dependencies only to run the tests.
```
Comment (irikuo) [inline _posts/2016-06-08-protractor-vs-intern-js.md]: It's `downside` (together). Also, Ruby has to start with capital R.

File: _posts/2016-06-08-protractor-vs-intern-js.md [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+  - More generic than Protractor (Integration testing should not be coupled with the application framework);
+  - Not very well documented (Documentation can be found [here](https://theintern.github.io/leadfoot/) but it's not very comprehensive. It does not describe very well how to use them and there are no practical usage examples);
+  - Not easy to find solutions when you get stuck;
+  - Lacks very basic functionality like condition find (After looking at the documentation and searching online I came up with a solution that worked but it looked quite messy);
+  - Syntax is hard to read and the nesting becomes hard to follow (hard to find where to put end() and where to put .then())
+  - Lack of basic methods like find element by text;
+  - It can be integrated easily with BrowserStack if you need to do cross browser checking. This is quite important if you need to test against a real browser and not phantom.js;
+  - For angular applications sometimes can be hard to debug (eg: to write in a text input you would expect that firstly you have to click the input to get the focus and then to type. Because of the working of angular this is flaky. Instead of this you can use type())
+
```
Comment (irikuo) [inline _posts/2016-06-08-protractor-vs-intern-js.md]: Check in all of this block the sentences are finished with a `;`. As this is a list of sentences, they should finish with a normal `.`. Check also names of AngularJS and PhantomJS and stuff like that, there are a couple of them to be fixed.

File: _posts/2016-06-08-protractor-vs-intern-js.md [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+  - Lack of basic methods like find element by text;
+  - It can be integrated easily with BrowserStack if you need to do cross browser checking. This is quite important if you need to test against a real browser and not phantom.js;
+  - For angular applications sometimes can be hard to debug (eg: to write in a text input you would expect that firstly you have to click the input to get the focus and then to type. Because of the working of angular this is flaky. Instead of this you can use type())
+  
+## Syntax
+
+    'it should allow the user to login': function () {
+        
+      return this.remote.setFindTimeout(5000)
+        .get('#/auth/sign_in').refresh()
```
Comment (irikuo) [inline _posts/2016-06-08-protractor-vs-intern-js.md]: Is that `refresh` needed for a simple example? The person reading the post might not know the context of what is being tested here, and be confused by this `refresh`, that if I recall correctly is needed to logout the user if there is a cookie set or something like that.

File: _posts/2016-06-08-protractor-vs-intern-js.md [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    },
+
+# Protractor
+  - Specific for angular applications:
+    - come with some methods specifically for angular which reduces the complexity of the specs (eg: by.model('person.name'), by.binding('person.concatName'), by.repeater('personhome.results'));
+    - wait to load is built in (in intern you have to specify for each test the default wait time with setFindTimeout());
+  - Intuitive syntax, easy to read and understand (specially if you come from a Ruby background);      
+  - Not a good choice for non angular applications;
+  - Easy to setup with Grunt
+  - More tutorials and practical example than Intern.js.
```
Comment (irikuo) [inline _posts/2016-06-08-protractor-vs-intern-js.md]: Same as in the Intern block, check some names (AngularJS, Intern...) and there are some `;` at the end of some sentences.

File: _posts/2016-06-08-protractor-vs-intern-js.md [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+        
+      var title = element(by.css('.side-menu'));
+      expect(title.getText()).toEqual('You are now logged in!');
+    });
+
+# Chosen Framework
+
+  Even if syntax is harder and not intuitive enough we went for intern as it would be easier to have the same framework to test all our applications. 
+  While you can use intern to do that it's not that easy to use Protractor to test a react application for example. The specific methods for protractor provides for angular applications are not a reason big enough to chose it over a more generic and better supported tool.
+
```
Comment (irikuo) [inline _posts/2016-06-08-protractor-vs-intern-js.md]: Same as before, review the names to match the official ones.

**Comment by irikuo** [comment]
I commented some lines to fix some stuff, like using the official names for libraries. On the other hand, feels like the list of Intern features are 80% bad points and just a couple of good things. Another point is that you talk about the confusing syntax, but maybe you can add that is using Promises to manage JavaScript's asynchronous nature. I am saying this because there are lots of people using promises and the syntax wouldn't be a problem for them.

The final conclusion seems weird, because looks like Intern is more generic but we don't like it and Protractor is much better but is Angular specific, so we chose Intern just because it's generic. I think if the reader doesn't know the context might find it hard to understand our choice. Maybe just by changing a bit the list of features of each library that feeling can go away.

File: _posts/2016-06-08-protractor-vs-intern-js.md [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+layout: post
+title:  Protractor or Intern ?
+date:   2016-06-08 14:00
+categories: blog
+navId: blog
+author: Alexandru Anca
+github: ionut998
+---
+
+The focus of this post is a comparison between Protractor and Intern for the integration testing of an AngularJS application.
```
Comment (richardbatty) [inline _posts/2016-06-08-protractor-vs-intern-js.md]: Slightly clunky intro. Maybe could say something like 'we've been trying to improve our use of integration tests - this post outlines how we chose between Protractor and Intern'

File: _posts/2016-06-08-protractor-vs-intern-js.md [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+
+# Current Implementation
+
+Currently the application is tested using Ruby and Capybara framework. As it turns out using this combination to test a single page application is very flaky. Out of 3 CI runs only 1 is usually successful, the rest is the result of flakiness.
```
Comment (richardbatty) [inline _posts/2016-06-08-protractor-vs-intern-js.md]: 'is tested using Ruby and Capybara framework' --> 'is tested using the Ruby-based Capybara framework'

Comment (richardbatty) [inline _posts/2016-06-08-protractor-vs-intern-js.md]: 'as it turns out' --> 'But' (it's more concise) or 'But we found that'

Comment (richardbatty) [inline _posts/2016-06-08-protractor-vs-intern-js.md]: 'Out of 3 CI runs only 1' --> 'Typically, out of 3 CI runs only 1' (without typically, it's not clear that you're talking about approximate numbers

Comment (richardbatty) [inline _posts/2016-06-08-protractor-vs-intern-js.md]: 'Typically, out of 3 CI runs only 1 is usually successful, the rest is the result of flakiness' -> 'Typically, out of 3 CI runs only 1 is successful, the rest is the result of flakiness'

File: _posts/2016-06-08-protractor-vs-intern-js.md [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+author: Alexandru Anca
+github: ionut998
+---
+
+The focus of this post is a comparison between Protractor and Intern for the integration testing of an AngularJS application.
+
+# Current Implementation
+
+Currently the application is tested using Ruby and Capybara framework. As it turns out using this combination to test a single page application is very flaky. Out of 3 CI runs only 1 is usually successful, the rest is the result of flakiness. 
+Another downside of this is that you need to install Ruby on top of your application dependencies only to run the tests.
```
Comment (richardbatty) [inline _posts/2016-06-08-protractor-vs-intern-js.md]: 'Another downside of this is that you need to install Ruby on top of your application dependencies only to run the tests.' --> 'Another downside of this is that you need to install Ruby on top of your application dependencies just to run the tests.'

File: _posts/2016-06-08-protractor-vs-intern-js.md [hunk 10]
```diff
@@ -0,0 +1,93 @@
...
+
+# Intern
+  - More generic than Protractor (Integration testing should not be coupled with the application framework).
```
Comment (richardbatty) [inline _posts/2016-06-08-protractor-vs-intern-js.md]: It's a little unusual to go straight into bullet points like this. Can you split it into two with 'Good points' and 'bad points'

File: _posts/2016-06-08-protractor-vs-intern-js.md [hunk 11]
```diff
@@ -0,0 +1,93 @@
...
+  - In Intern everything is a promise. This is the way intern handles asynchronicity. The method `setFindTimeout()` sets the default wait time for elements to be loaded. If the page will take longer than specified `setFindTimeout()` to load then the test will fail.
```
Comment (richardbatty) [inline _posts/2016-06-08-protractor-vs-intern-js.md]: 'If the page will take longer than specified `setFindTimeout()` to load then the test will fail.' 

I don't understand what you mean, is this it?: 'If the page takes longer to load than specified by `setFindTimeout()` then the test will fail.'

File: _posts/2016-06-08-protractor-vs-intern-js.md [hunk 12]
```diff
@@ -0,0 +1,93 @@
...
+  - It can be integrated easily with BrowserStack if you need to do cross browser checking. This is quite important if you need to test against a real browser and not PhantomJS.
```
Comment (richardbatty) [inline _posts/2016-06-08-protractor-vs-intern-js.md]: 'cross-browser'

delete 'quite' to be more concise.

'real browser and not PhantomJS' --> 'real browser instead of PhantomJS'

File: _posts/2016-06-08-protractor-vs-intern-js.md [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+  - Not very well documented (Documentation can be found [here](https://theintern.github.io/leadfoot/) but it's not very comprehensive. It does not describe very well how to use them and there are no practical usage examples).
```
Comment (richardbatty) [inline _posts/2016-06-08-protractor-vs-intern-js.md]: 'It does not have good descriptions of how to use it and there are no practical examples'

File: _posts/2016-06-08-protractor-vs-intern-js.md [hunk 14]
```diff
@@ -0,0 +1,93 @@
...
+  - Not easy to find solutions when you get stuck.
```
Comment (richardbatty) [inline _posts/2016-06-08-protractor-vs-intern-js.md]: Do you mean that because not many people use it there aren't good blog posts and stack overflow answers?

File: _posts/2016-06-08-protractor-vs-intern-js.md [hunk 15]
```diff
@@ -0,0 +1,93 @@
...
+  - Lacks very basic functionality like find element by text or condition find (After looking at the documentation and searching online I came up with a solution that worked but it looked quite messy).
```
Comment (richardbatty) [inline _posts/2016-06-08-protractor-vs-intern-js.md]: not sure what you mean by 'condition find'. If you mean find depending on some variable, I guess you'd call this 'conditional find'

File: _posts/2016-06-08-protractor-vs-intern-js.md [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+  - Syntax is hard to read and the nesting becomes hard to follow (hard to find where to put `.end()` and where to resolve the promise with `.then()`).
```
Comment (richardbatty) [inline _posts/2016-06-08-protractor-vs-intern-js.md]: '(for example, it's hard to find ...'

File: _posts/2016-06-08-protractor-vs-intern-js.md [hunk 17]
```diff
@@ -0,0 +1,93 @@
...
+  - It can be integrated easily with BrowserStack if you need to do cross browser checking. This is quite important if you need to test against a real browser and not PhantomJS.
+  - Not very well documented (Documentation can be found [here](https://theintern.github.io/leadfoot/) but it's not very comprehensive. It does not describe very well how to use them and there are no practical usage examples).
+  - Not easy to find solutions when you get stuck.
+  - Lacks very basic functionality like find element by text or condition find (After looking at the documentation and searching online I came up with a solution that worked but it looked quite messy).
+  - Syntax is hard to read and the nesting becomes hard to follow (hard to find where to put `.end()` and where to resolve the promise with `.then()`).  
+  - For AngularJS applications sometimes can be hard to debug (eg: to write in a text input you would expect that firstly you have to click the input to get the focus and then to type. Because of the working of AngularJS this is flaky. Instead of this you can use `type()`).
```
Comment (richardbatty) [inline _posts/2016-06-08-protractor-vs-intern-js.md]: 'For AngularJS applications sometimes can be hard to debug' --> 'It can be hard to debug intern tests running against an AngularJS application'

e.g. doesn't need a colon, it's usually used like 'e.g. to write in a text ...'

File: _posts/2016-06-08-protractor-vs-intern-js.md [hunk 18]
```diff
@@ -0,0 +1,93 @@
...
+   
+# Protractor
+  - Specific for AngularJS applications:
+    - Come with some methods specifically for AngularJS which reduces the complexity of the specs (eg: `by.model('person.name')`,` by.binding('person.concatName')`, `by.repeater('person.results')`).
```
Comment (richardbatty) [inline _posts/2016-06-08-protractor-vs-intern-js.md]: 'come with' -> 'It comes with'

eg: -> e.g.

File: _posts/2016-06-08-protractor-vs-intern-js.md [hunk 19]
```diff
@@ -0,0 +1,93 @@
...
+    - To solve the asynchronicity problem Protractor binds to AngularJS elements to check when the elements finished loading.
```
Comment (richardbatty) [inline _posts/2016-06-08-protractor-vs-intern-js.md]: 'elements finished' -> 'elements have finished'

File: _posts/2016-06-08-protractor-vs-intern-js.md [hunk 20]
```diff
@@ -0,0 +1,93 @@
...
+    - Easy to learn if you know AngularJS
+  - Intuitive syntax, easy to read and understand (specially if you come from a Ruby background).
+  - Not a good choice for non AngularJS applications. If used for non AngularJS applications there is no way of ensuring that elements are loaded dynamically therefore you will have to use `setTimeout()`.
```
Comment (richardbatty) [inline _posts/2016-06-08-protractor-vs-intern-js.md]: non AngularJS -> non-AngularJS

Comment (richardbatty) [inline _posts/2016-06-08-protractor-vs-intern-js.md]: 'loaded dynamically therefore' -> 'loaded dynamically, so'

File: _posts/2016-06-08-protractor-vs-intern-js.md [hunk 21]
```diff
@@ -0,0 +1,93 @@
...
+  element(by.css('[type="submit"]')).click();
+  // check login message
+  var title = element(by.css('.side-menu'));
+  expect(title.getText()).toEqual('You are now logged in!');
+});
+{% endhighlight %}
+
+# Chosen Framework
+
+By looking at the pros and cons of both frameworks we can see more pros for Protractor therefore seem like the right choice to make. That is only true for AngularJS applications. If you have to test a non AngularJS application worth protractor you will need a way to deal with the asynchronicity and you can end up with some flaky tests.
```
Comment (richardbatty) [inline _posts/2016-06-08-protractor-vs-intern-js.md]: for Protractor, so it seems like the right choice to make

File: _posts/2016-06-08-protractor-vs-intern-js.md (hunk 12 above)
Comment (errm) [inline _posts/2016-06-08-protractor-vs-intern-js.md]: Is this an advantage over and above what we could do we Protractor or Capybara, surely they both are supported by browser stack too?

File: _posts/2016-06-08-protractor-vs-intern-js.md (hunk 10 above)
Comment (errm) [inline _posts/2016-06-08-protractor-vs-intern-js.md]: Is there a reason we wouldn't want to couple with the framework?
It sounds like a noble goal, but practically is there any advantage... I imagine that the tests have to become somewhat coupled to the way the framework renders stuff anyway, just to work?  Is there any real possibility of integration tests in this scenario still working were we to switch to another framework?

**Comment by ionut998** [comment]
I have done the suggested changes. Thanks @irikuo and @richardbatty for help
